### PR TITLE
[Execution] Use generic slice algorithms instead of interfaces

### PR DESCRIPTION
**Pull request by SaveTheRbtz**
Generic version is faster than the interface-based one.  While here, also:
* Simplify `RegisterUpdates` -- no need to roundtrip data through `RegisterEntries` just for sorting.
(Followup from #3529)
* Simplify `Union` -- a roundtrip through map is negligible compared to O(n*log(n)) sorting.
* Simplify the rest of `IdentityList` methods to remove the custom binary searches and equality functions.

While here, also mark a race in `DeterministicSample`. 

This also replaces custom implementation of byte comparisons with a standard `Bytes.Compare()` that supports both AVX and [AVX2 on x86_64](https://cs.opensource.google/go/go/+/refs/tags/go1.19.3:src/internal/bytealg/compare_amd64.s;l=194), hence using `VPCMPEQB` to do 64-byte compares.

Side note about `IdentityList`: looking at the code, it should really be a `IdentityOrderedSet` and internally be represented by a `map[Identiy]struct{}` and a list for keeping track of the order. This is outside of the scope of this diff though.

### CODE REVIEW

**Review by pattyshack — APPROVED** [review]

**Review by AlexHentschel — APPROVED** [review]

File: model/flow/identity.go [hunk 1]
```diff
@@ -591,66 +541,20 @@ func (il IdentityList) Exists(target *Identity) bool {
...
-	ok := false
-	for i, id := range il.NodeIDs() {
-		if id == identifier {
-			index = i
-			ok = true
-			break
-		}
+func (il IdentityList) GetIndex(target Identifier) (uint, bool) {
+	i := slices.IndexFunc(il, func(a *Identity) bool {
+		return bytes.Equal(a.NodeID[:], target[:])
```
Comment (AlexHentschel) [inline model/flow/identity.go:554]: `Identifier` is a fixed-length array, which can be directly compared in go without the need to convert to slices:
```suggestion
		return a.NodeID == target
```

File: model/flow/identity.go [hunk 2]
```diff
@@ -466,6 +458,7 @@ func (il IdentityList) Sample(size uint) IdentityList {
 }
 
 // DeterministicSample returns deterministic random sample from the `IdentityList` using the given seed
+// TODO(rbtz): fix race condition
 func (il IdentityList) DeterministicSample(size uint, seed int64) IdentityList {
```
[the remark this answers is not in the document]
Comment (SaveTheRbtz) [reply model/flow/identity.go:462]: yep, fix is rather simple, I just did not want to couple it with this diff -- it is big as it is and `DeterministicSample` is not used anywhere.

File: fvm/state/state.go [hunk 3]
```diff
@@ -282,23 +282,17 @@ func (s *State) MergeState(other *State) error {
...
-	i := 0
 	for k := range s.updatedAddresses {
-		addresses[i] = k
-		i++
+		addresses = append(addresses, k)
 	}
 
-	sort.Sort(addresses)
+	slices.SortFunc(addresses, func(a, b flow.Address) bool {
+		return bytes.Compare(a[:], b[:]) < 0
```
[the remark this answers is not in the document]
Comment (SaveTheRbtz) [reply fvm/state/state.go]: Nice catch!

File: model/flow/identity.go [hunk 4]
```diff
@@ -507,76 +500,33 @@ func (il IdentityList) SamplePct(pct float64) IdentityList {
...
-		if union[lastUnique].NodeID != union[i].NodeID {
-			lastUnique++
-			union[lastUnique] = union[i]
-		}
 	}
-	return union[:lastUnique+1]
+
+	slices.SortFunc(union, func(a, b *Identity) bool {
+		return bytes.Compare(a.NodeID[:], b.NodeID[:]) < 0
+	})
```
[the remark this answers is not in the document]
Comment (SaveTheRbtz) [reply model/flow/identity.go:519]: The logic behind this is:
* if list is big then O(n*log(n)) of sorting will dominate any linear allocation costs.
* if lists are small then the allocation do not matter.

Given that Union is mostly unused outside of tests, it does look like a premature optimization.

**Comment by SaveTheRbtz** [comment]
bors merge

File: model/flow/identity.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (SaveTheRbtz) [reply model/flow/identity.go:462]: https://github.com/onflow/flow-go/pull/3559

**Pull request merged**